validator/coordinator/src/main/java/tech/pegasys/artemis/validator/coordinator/KeystoresValidatorKeyProvider.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.validator.coordinator;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static java.lang.String.format;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.stream.Collectors.toMap;
+import static org.apache.commons.lang3.StringUtils.isEmpty;
+import static tech.pegasys.artemis.util.alogger.ALogger.STDOUT;
+
+import com.google.common.io.Files;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.logging.log4j.Level;
+import org.apache.tuweni.bytes.Bytes;
+import tech.pegasys.artemis.bls.keystore.KeyStore;
+import tech.pegasys.artemis.bls.keystore.KeyStoreLoader;
+import tech.pegasys.artemis.bls.keystore.KeyStoreValidationException;
+import tech.pegasys.artemis.bls.keystore.model.KeyStoreData;
+import tech.pegasys.artemis.util.bls.BLSKeyPair;
+import tech.pegasys.artemis.util.config.ArtemisConfiguration;
+import tech.pegasys.artemis.util.mikuli.KeyPair;
+import tech.pegasys.artemis.util.mikuli.SecretKey;
+
+public class KeystoresValidatorKeyProvider implements ValidatorKeyProvider {
+  static final int KEY_LENGTH = 48;
+
+  @Override
+  public List<BLSKeyPair> loadValidatorKeys(final ArtemisConfiguration config) {
+    final List<Pair<Path, Path>> keystorePasswordFilePairs =
+        config.getValidatorKeystorePasswordFilePairs();
+    checkNotNull(keystorePasswordFilePairs, "validator keystore and password pairs cannot be null");
+
+    // return distinct loaded key pairs
+
+    return new ArrayList<>(
+        keystorePasswordFilePairs.stream()
+            .map(
+                pair -> {
+                  final String password = loadPassword(pair.getRight());
+                  final Bytes privKey = loadBLSPrivateKey(pair.getLeft(), password);
+                  return new BLSKeyPair(new KeyPair(SecretKey.fromBytes(padLeft(privKey))));
+                })
+            .collect(
+                toMap(
+                    blsKeyPair -> blsKeyPair.getPublicKey().toString(),
+                    Function.identity(),
+                    (dupKey1, dupKey2) -> dupKey2))
+            .values());
```

Comment:
You can just use:
```suggestion
     return keystorePasswordFilePairs.stream()
        .map(pair -> padLeft(loadBLSPrivateKey(pair.getLeft(), loadPassword(pair.getRight()))))
        .distinct()
        .map(privKey -> new BLSKeyPair(new KeyPair(SecretKey.fromBytes(privKey))))
        .collect(toList());
```

Reply:
very clever, i forgot we have distinct in streams api as well.

Reply:
in fact this has taken care of the reason why I used publickey as map key because BLsKeyPair was always getting multiple instances even with same private/public key (because of raw bytes comparison somewhere deep within).